### PR TITLE
Upgrading Gloo

### PR DESCRIPTION
**Pull request by rickducott**

### CODE REVIEW

File: gloo/docs/upgrading/0.18.0.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+# Upgrading to 0.18.0 on Kubernetes
+In Gloo 0.18.0, we updated the Gateway API to support managing Envoy TCP (L4) routing configuration in addition to existing HTTP/S (L7). This required making a non-backwards compatible change to the Gateway CRD in Gloo. 
+
+Due to limited support for CRD versioning in older versions of Kubernetes (we cannot assume our customers run Kubernetes 1.13+), we implemented this change with a new Gateway v2 CRD. Gloo 0.18.0+ will no longer install the old Gateway CRD or controller.  
+
+This guide documents a process for safely upgrading Gloo to 0.18.0, with an emphasis on minimizing manual work and avoiding downtime while shifting traffic to the new Gateway implementation. 
+
+## Upgrade Steps
+### 1. Install Gloo v0.18.0 with the `upgrade` flag
+If installing with `glooctl`, provide the `--upgrade` flag. If installing with Helm, set `upgrade: true` in the values file.
```
Comment (joekelley) [inline gloo/docs/upgrading/0.18.0.md]: as implemented, this is currently gateway.upgrade.

File: gloo/docs/upgrading/0.18.0.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+Once you are satisfied with traffic on Gateway v2, you can edit the `gateway-proxy` service to select on the following 2 labels exlusively.
+```yaml
+selector:
+	gloo: gateway-proxy-v2
+	gateway-proxy: live
+```
+
+Verify that all traffic is successfully routed to `gateway-proxy-v2`.
+
+### 4. Delete old Gateway v1 resources
```
Comment (joekelley) [inline gloo/docs/upgrading/0.18.0.md]: we can also delete the job and existing proxies

File: gloo/docs/upgrading/0.18.0.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+The `gateway-proxy-v2` configuration can be tested. 
+* Run `glooctl get proxy` and verify that the `gateway-proxy-v2` status is `ACCEPTED`. 
+* Test traffic against `gateway-proxy-v2` to ensure it behaves similar to `gateway-proxy`
+	This can be done with port-forwarding: `kubectl port-forward -n gloo-system deployment/gateway-proxy-v2 8080`
+	Or by setting up a new service / ingress manually for testing. 
+
+### 4. Migrate traffic to Gateway v2
+Prior to upgrading, all traffic was being routed through the Envoy instance inside `gateway-proxy`. Now we can migrate traffic from `gateway-proxy` to `gateway-proxy-v2`. 
+
+In order to start shifting 50% of the traffic to the new `gateway-proxy-v2`. The `gateway-proxy: live` label needs to be added to the `gateway-proxy-v2` deployment.
```
Comment (rickducott) [inline gloo/docs/upgrading/0.18.0.md]:
```suggestion
In order to start shifting 50% of the traffic to the new `gateway-proxy-v2`, the `gateway-proxy: live` label needs to be added to the `gateway-proxy-v2` deployment.
```

File: gloo/docs/upgrading/0.18.0.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+We recommend that you update your load balancer to handle routing traffic to two services: `gateway-proxy` and the new `gateway-proxy-v2`. If that is not possible in your environment, you can follow the steps below. However, installation for subsequent versions of Gloo may assume that you are running `gateway-proxy-v2` exclusively.
+
+## Upgrade Steps
+
+This guide documents a process for safely upgrading Gloo to 0.18.0, with an emphasis on minimizing manual work and avoiding downtime while shifting traffic to the new Gateway implementation. 
+
+### 1. Prepare your existing resources for upgrade
+To facilitate future updates, we've introduced live labels to our gateway-proxy deployments and services. You can take advantage of this feature during the migration to v0.18.0 by updating your existing resources. Run `kubectl edit deployment -n gloo-system gateway-proxy` and add `gateway-proxy: live` to the labels map under spec.template.metadata. Similarly, run `kubectl edit service -n gloo-system gateway-proxy`, add `gateway-proxy: live` to the selector map under spec, and remove all other selectors. Your gateway should function as before.
```
Comment (EItanya) [inline gloo/docs/upgrading/0.18.0.md]: I think it might be worth making it more clear that without the new label the transition will not work seamlessly

Comment (joekelley) [reply gloo/docs/upgrading/0.18.0.md]: done added a line about that

File: gloo/docs/upgrading/0.18.0.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+Adding this label and selector will make it possible to route traffic through `gateway-proxy-v2` from the `gateway-proxy`, minimizing downtime.
```
Comment (rickducott) [inline gloo/docs/upgrading/0.18.0.md]: What does this sentence mean?

File: gloo/docs/upgrading/0.18.0.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+### 2. Install Gloo v0.18.0 with the `upgrade` flag
+If installing with `glooctl`, provide the `--upgrade` flag. If installing with Helm, set `upgrade: true` in the values file under `gateway`.
```
Comment (rickducott) [inline gloo/docs/upgrading/0.18.0.md]: Let's document this in the helm values table

File: gloo/docs/upgrading/0.18.0.md (hunk 5 above)
Comment (joekelley) [reply gloo/docs/upgrading/0.18.0.md]: Oops, added the word "service", let me know if it's still unclear.

File: gloo/docs/installation/gateway/kubernetes/_index.md [hunk 7]
```diff
@@ -121,6 +121,7 @@ The table below describes all the values that you can override in your custom va
...
 | gateway.enabled                                           | bool     | enable Gloo API Gateway features                                                                                                                                                                                                                               |
+| gateway.upgarde                                           | bool     | Deploy a Job to convert (but not delete) v1 Gateway resources to v2 and not add a "live" label to the gateway-proxy deployment's pod template. This allows for canary testing of gateway-v2 alongside an existing instance of gloo running with v1 gateway resources and controllers.                                                                                                                                                                                                                              |
```
Comment (rickducott) [inline gloo/docs/installation/gateway/kubernetes/_index.md]: upgrade

Comment (rickducott) [inline gloo/docs/installation/gateway/kubernetes/_index.md]:
```suggestion
| gateway.upgrade                                           | bool     | Deploy a Job to convert (but not delete) v1 Gateway resources to v2 and not add a "live" label to the gateway-proxy deployment's pod template. This allows for canary testing of gateway-v2 alongside an existing instance of gloo running with v1 gateway resources and controllers.                                                                                                                                                                                                                              |
```

**Pull request merged**